### PR TITLE
sc-2163 test case for successful Admin API authentication

### PR DESCRIPTION
**Pull request by pdeziel**
This adds the missing test cases for successful authentication on the Admin API `Authenticate` endpoint. The Admin API handler uses Google's `idtoken.Validate()` and is expecting a Google token. I got around this by adding a `validate` function to the token manager which must be configured when a new `TokenManager` is instantiated. When we're running the tests, the mocked `TokenManager` will be used which uses a mocked validation function instead of the Google one. The mocked function simply parses the token and injects the claims into the returned payload, bypassing all the checks that the Google handler is doing. This allows us to test that we are handling authentication properly assuming that we were passed a valid credential.

### CODE REVIEW

File: pkg/gds/admin.go [hunk 1]
```diff
@@ -247,7 +247,7 @@ func (s *Admin) Authenticate(c *gin.Context) {
 	}
 
 	// Validate the credential with Google
-	if claims, err = idtoken.Validate(c.Request.Context(), in.Credential, s.conf.Oauth.GoogleAudience); err != nil {
+	if claims, err = s.tokens.Validate(c.Request.Context(), in.Credential, s.conf.Oauth.GoogleAudience); err != nil {
```
Comment (rebeccabilbro) [inline pkg/gds/admin.go:250]: Guessing this was just a stray bug you discovered while testing?

File: pkg/gds/mock.go [hunk 2]
```diff
@@ -88,7 +88,7 @@ func MockConfig() config.Config {
 			CookieDomain: "admin.gds.dev",
 			Audience:     "http://api.admin.gds.dev",
 			Oauth: config.OauthConfig{
-				GoogleAudience:         "4284607864536-notarealgoogleaudience.apps.googleusercontent.com",
+				GoogleAudience:         "http://localhost",
```
Comment (rebeccabilbro) [inline pkg/gds/mock.go:91]: Changed to make the integration testing easier?

File: pkg/gds/tokens/mock.go [hunk 3]
```diff
@@ -25,3 +30,40 @@ func MockTokenManager() (tm *TokenManager, err error) {
 	tm.keys[tm.currentKeyID] = &key.PublicKey
 	return tm, nil
 }
+
+// mockValidate is a mock for Google's token validation which enables authentication
+// testing by internally parsing the given token and injecting the claims into the
+// returned payload. This will only validate badly constructed tokens and tokens where
+// the supplied audience does not match the audience used to sign the token.
```
Comment (rebeccabilbro) [inline pkg/gds/tokens/mock.go:37]: You mention that this mock will only allow us to catch bad tokens and mismatched audiences; are there other edge cases we should be thinking about moving forward?

**Review by rebeccabilbro — APPROVED** [review]
This looks good to me @pdeziel! Abstracting out the validation mechanism to make it easier to test seems to have worked nicely -- I left you a few questions/comments (not as familiar with the Admin API so just trying to get up to speed), but I think this should be good to merge 👍

File: pkg/gds/mock.go (hunk 2 above)
Comment (pdeziel) [reply pkg/gds/mock.go:91]: The other option was to change the secrets mock to use the fake google audience above, for testing purposes they just have to match so I went for the simpler option.

File: pkg/gds/tokens/mock.go (hunk 3 above)
Comment (pdeziel) [reply pkg/gds/tokens/mock.go:37]: Good question, we could also verify that the expiration time and issued time make sense which isn't that complicated of a check. Currently we're passing basically all of the token validation off to Google so from a testing perspective there's only really one validation error path, so I didn't find it absolutely necessary to include those checks in the mock for the scope of what this PR is testing.
